### ISSUE TITLE
IterableMappingMethod uses List generated method for a Collection if the list is defined before the collection

### ISSUE THREAD

**Issue by filiphr** [mapstruct/mapstruct#971]
Example:

```java
public class Target {
    List<Integer> integers;
    Collection<Integer> integersCollection;

   //The order of the getters is not really important here
}

public class Source {
    List<String> integers;
    List<String> integersCollection;

   //getters for the integers are before the getters of the integersCollection
}

@Mapper
public interface SourceTargetMapper {
    Source targetToSource(Target target);
}
```

The Implementation:
```java

public class CollectionMapperImpl implements CollectionMapper {

    @Override
    public Source targetToSource(Target target) {
        if ( target == null ) {
            return null;
        }

        Source source = new Source();

        List<String> list = integerListToStringList( target.getIntegers() );
        if ( list != null ) {
            source.setStringStream( list );
        }
        List<String> list_ = integerListToStringList( target.getIntegersCollection() );
        if ( list_ != null ) {
            source.setStringCollection( list_ );
        }

        return source;
    }

    protected List<String> integerListToStringList(List<Integer> list) {
        if ( list == null ) {
            return null;
        }

        List<String> list_ = new ArrayList<String>();
        for ( Integer integer : list ) {
            list_.add( String.valueOf( integer ) );
        }

        return list_;
    }
}
```

If the order of the getters is inversed in the `Source` then it works the first forged method is:
```java
protected List<String> integerCollectionToStringList(Collection<Integer> collection) {
        if ( collection == null ) {
            return null;
        }

        List<String> list = new ArrayList<String>();
        for ( Integer integer : collection ) {
            list.add( String.valueOf( integer ) );
        }

        return list;
}
```

And this one works both for the `List<Integer>` and the `Collection<Integer>`.

One solution of this would be to check if the `sourceParameters` in the `IterableMappingMethod` are equal, but this would always generate more specific methods (one for `List`, and another one for `Collection`), or do something more clever.

**Comment by filiphr** [comment]
The same happens during the mapping of different `Map`s.

**Comment by filiphr** [comment]
Fixed with PR #972
